### PR TITLE
ENH: Imprpve interval_range and IntervalIndex

### PR DESCRIPTION
**Pull request by bashtage**
- [x] Tests added: Please use `assert_type()` to assert the type of any return value

### CODE REVIEW

File: pandas-stubs/core/indexes/interval.pyi [hunk 1]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
         name: Hashable = ...,
         copy: bool = ...,
         dtype: IntervalDtype | None = ...,
     ) -> IntervalIndex: ...
+    def __contains__(self, key: Any) -> bool: ...
     def astype(self, dtype: DtypeArg, copy: bool = ...) -> IntervalIndex: ...
     @property
     def inferred_type(self) -> str: ...
     def memory_usage(self, deep: bool = ...) -> int: ...
+    def is_monotonic_decreasing(self) -> bool: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:86]: This is a `property` of `Index`, so I think you can delete this here.  The implementation is different, but we can just inherit from `Index`

File: pandas-stubs/core/indexes/interval.pyi [hunk 2]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
+    def is_unique(self) -> bool: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:87]: inherited from `Index`, so delete

File: pandas-stubs/core/indexes/interval.pyi [hunk 3]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
     @property
     def is_overlapping(self) -> bool: ...
-    def get_loc(self, key, tolerance=...) -> int | slice | np.ndarray: ...
+    # Note: tolerance removed as it has no effect
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:90]: Should the pandas docs remove it from the signature?

File: pandas-stubs/core/indexes/interval.pyi [hunk 4]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
+    def get_loc(
+        self,
+        key: Label,
+        method: str | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:94]: `FillnaOptions | Literal['nearest'] | None`

Could fix it in `core/indexes/base.pyi` as well.

File: pandas-stubs/core/indexes/interval.pyi [hunk 5]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
+    ) -> int | slice | np.ndarray: ...
     def get_indexer(
         self,
-        targetArrayLike,
+        target: Index,
         method: str | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:99]: same comment on `method` being specific `Literal` values

File: pandas-stubs/core/indexes/interval.pyi [hunk 6]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
         limit: int | None = ...,
         tolerance=...,
     ) -> np.ndarray: ...
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:102]: `np.NDArray[np.intp]`

File: pandas-stubs/core/indexes/interval.pyi [hunk 7]
```diff
@@ -47,37 +69,57 @@ class IntervalIndex(IntervalMixin, ExtensionIndex):
...
-    def __gt__(self, other): ...
-    def __ge__(self, other): ...
 
 def interval_range(
-    start=..., end=..., periods=..., freq=..., name=..., closed: str = ...
-): ...
+    start: int | float | DatetimeLike | None = ...,
+    end: int | float | DatetimeLike | None = ...,
+    periods: int | None = ...,
+    freq: int | str | BaseOffset | None = ...,
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:122]: docs say `DateOffset` here, not `BaseOffset`

**Review by Dr-Irv — CHANGES_REQUESTED** [review]

File: pandas-stubs/core/indexes/interval.pyi (hunk 1 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Removed.

File: pandas-stubs/core/indexes/interval.pyi (hunk 3 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: I think it has to stay due to inheritance.  I changed the comment.

File: pandas-stubs/core/indexes/interval.pyi (hunk 4 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Fixed in both.

File: pandas-stubs/core/indexes/interval.pyi (hunk 5 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Fixed.

File: pandas-stubs/core/indexes/interval.pyi (hunk 6 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Fixed.

File: pandas-stubs/core/indexes/interval.pyi (hunk 7 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Fixed.

File: pandas-stubs/core/indexes/interval.pyi (hunk 2 above)
Comment (bashtage) [reply pandas-stubs/core/indexes/interval.pyi]: Removed.

**Comment by bashtage** [comment]
Tests added, comments fixed, and green.

File: pandas-stubs/core/indexes/interval.pyi [hunk 8]
```diff
@@ -25,6 +31,20 @@ from pandas._typing import (
 from pandas.core.dtypes.dtypes import IntervalDtype as IntervalDtype
 from pandas.core.dtypes.generic import ABCSeries
 
+_Edges: TypeAlias = Union[
```
Comment (Dr-Irv) [inline pandas-stubs/core/indexes/interval.pyi:34]: nice idea

**Review by Dr-Irv — APPROVED** [review]
thanks @bashtage

File: pandas-stubs/_typing.pyi [hunk 9]
```diff
@@ -44,9 +44,7 @@ PandasScalar: TypeAlias = Union[
 ]
 # Scalar: TypeAlias = Union[PythonScalar, PandasScalar]
 
-DatetimeLike: TypeAlias = Union[
-    datetime.date, datetime.datetime, np.datetime64, Timestamp
-]
+DatetimeLike: TypeAlias = Union[datetime.datetime, np.datetime64, Timestamp]
```
Comment (gandhis1) [inline pandas-stubs/_typing.pyi:47]: It appears that `date` is incompatible with this change. `datetime` is a sub class of `date`, so the opposite change - removing `datetime`, would probably work.

Comment (bashtage) [reply pandas-stubs/_typing.pyi:47]: Explainer for the narrowing is in #369

**Pull request merged**